### PR TITLE
add guestsfilelist

### PR DESCRIPTION
**Pull request by cortho**
for #87, functionality as in files_sharing. I hope I didn't miss something

who could review this?

### CODE REVIEW

**Comment by pmaier1** [comment]
@davitol or @PVince81 to review or delegate pls

File: controller/guestsharecontroller.php [hunk 1]
```diff
@@ -0,0 +1,229 @@
+<?php
+/**
+ * @author Michael Jobst <mjobst+github@tecratech.de>
```
Comment (PVince81) [inline controller/guestsharecontroller.php]: if you didn't copy it 1 to 1, please adjust authors

File: css/guests.css [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.nav-icon-sharingguests {
+	background-image: url('../../files/img/share.svg');
```
Comment (PVince81) [inline css/guests.css]: Please reuse whatever class is used in core. I'm not a fan of having one app reference the files of another app like this.

If core is missing a proper CSS, please submit a PR to core to make that possible. Thanks

File: js/guestsfilelist.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2014 Vincent Petry <pvince81@owncloud.com>
```
Comment (PVince81) [inline js/guestsfilelist.js]: adjust this ?

File: js/guestsfilelist.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			{
+				id: 'shares.guests',
+				scrollContainer: $('#app-content'),
+				fileActions: this._createFileActions(),
+				config: OCA.Files.App.getFilesConfig()
+			}
+		);
+
+		this._extendFileList(this._guestsFileList);
+		this._guestsFileList.appName = t('guests', 'Shared with Guests');
```
Comment (PVince81) [inline js/guestsfilelist.js]: @pmaier1 capital "Guests" or "guests" ?

File: js/sharedfilelist.js [hunk 5]
```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2014 Vincent Petry <pvince81@owncloud.com>
```
Comment (PVince81) [inline js/sharedfilelist.js]: please adjust

File: js/sharedfilelist.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+ * This file is licensed under the Affero General Public License version 3
+ * or later.
+ *
+ * See the COPYING-README file.
+ *
+ */
+(function() {
+
+	/**
+	 * @class OCA.Sharing.FileList
```
Comment (PVince81) [inline js/sharedfilelist.js]: adjust namespace

File: js/sharedfilelist.js [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+	 *
+	 * @param $el container element with existing markup for the #controls
+	 * and a table
+	 * @param [options] map of options, see other parameters
+	 */
+	var FileList = function($el, options) {
+		this.initialize($el, options);
+	};
+	FileList.prototype = _.extend({}, OCA.Files.FileList.prototype,
+		/** @lends OCA.Sharing.FileList.prototype */ {
```
Comment (PVince81) [inline js/sharedfilelist.js]: adjust namespace

File: js/sharedfilelist.js [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+			return true;
+		},
+
+		/**
+		 * Converts the OCS API share response data to a file info
+		 * list
+		 * @param {Array} data OCS API share array
+		 * @return {Array.<OCA.Sharing.SharedFileInfo>} array of shared file info
+		 */
+		_makeFilesFromShares: function(data) {
```
Comment (PVince81) [inline js/sharedfilelist.js]: I wonder if you could simply extend `OCA.Sharing.FileList` which already contains this code, and basically just change the `reload` function to point at your guest specific endpoint ?

File: list.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<?php
+/**
+ * @author Vincent Petry <pvince81@owncloud.com>
```
Comment (PVince81) [inline list.php]: please adjust

File: list.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+// Check if we are a user
+OCP\User::checkLoggedIn();
+
+$tmpl = new OCP\Template('guests', 'list', '');
+
+OCP\Util::addScript('guests', 'guestsfilelist');
```
Comment (PVince81) [inline list.php]: this file contains a class called `App`, I suggest to rename the file to "app.js" to avoid confusion and rename your "sharedfilelist" to "guestsfilelist". I was confused when seeing both and had to look it up. Future devs will likely be confused too.

File: js/guestsfilelist.js (hunk 4 above)
Comment (pmaier1) [inline js/guestsfilelist.js]: Lower case, please, to stay in consistency with the other filters.

**Comment by cortho** [comment]
@PVince81 Thanks a lot, I absolutely missed that we can extend OCA.Sharing.FileList

**Comment by felixboehm** [comment]
@PVince81 needed for review

**Comment by PVince81** [comment]
@felixboehm review already done, changes are required.

> What exactly do you mean by that, you could elaborate on it?

@cortho 

1. Share text file with guest
1. Go to guest file list
1. Click text file

Expected: text editor opens

**Comment by cortho** [comment]
@PVince81 I'm still not sure, which kind of tests you mean exactly, but when you are writing 'click', I guess this means UI Tests like https://doc.owncloud.org/server/10.0/developer_manual/core/ui-testing.html
If this is correct I'll read into selenium tests and try to get this done, if not please tell me which test suite is the most suitable.

**Comment by PVince81** [comment]
I was thinking manual testing. But if we can have selenium tests that would be even better.

@individual-it or @phil-davis could help set this up

**Comment by phil-davis** [comment]
@cortho I will guess that @PVince81 just means to manually make sure the UI works.
So far there are no automated selenium UI tests that drive the UI to create a new file/folder... but that is definitely something that is needed some time.
These days, with a core-only system, there is no files_texeditor app installed. So on the files page, when you click the "+" you only get the options to upload a file and to create a folder. So an automated test that creates a text file using the UI would have to go in a UI test suite in files_texteditor.
And then thought needs to go into the architecture for running tests that need multiple apps - e.g. being a guest with guests app and creating files with files_xxxeditor. Where does the test feature live, how does it make sure that the right mix of apps is installed...

Just having a brain dump here about the future of this sort of testing.!

**Comment by cortho** [comment]
@phil-davis thanks a lot, I'll watch the testing implementations so I can learn them on the fly

@PVince81 with manual testing this PR works as expected. (i.e. behavior is the same as for 'Shared with me/others'), so I can open any files as expected, this includes that the text editor is opening, when I click a text file, that is shared with a guest.

**Comment by PVince81** [comment]
@cortho ok cool. Let us know when you are done with the review changes.

File: controller/guestsharecontroller.php (hunk 1 above)
Comment (cortho) [inline controller/guestsharecontroller.php]: adjusted

File: css/guests.css (hunk 2 above)
Comment (cortho) [inline css/guests.css]: took css from core

File: js/guestsfilelist.js (hunk 3 above)
Comment (cortho) [inline js/guestsfilelist.js]: adjusted

File: js/guestsfilelist.js (hunk 4 above)
Comment (cortho) [inline js/guestsfilelist.js]: done: lower case

File: js/sharedfilelist.js (hunk 5 above)
Comment (cortho) [inline js/sharedfilelist.js]: adjusted

File: js/sharedfilelist.js (hunk 6 above)
Comment (cortho) [inline js/sharedfilelist.js]: adjusted

File: js/sharedfilelist.js (hunk 7 above)
Comment (cortho) [inline js/sharedfilelist.js]: adjusted

File: js/sharedfilelist.js (hunk 8 above)
Comment (cortho) [inline js/sharedfilelist.js]: ack. done

File: list.php (hunk 9 above)
Comment (cortho) [inline list.php]: adjusted

**Comment by cortho** [comment]
@PVince81 I performed all of your change request with my last commit. Did I miss some github functionality to have you informed?

**Comment by PVince81** [comment]
Not sure how to trigger a "I made the changes" in Github. Usually simply post a comment "Please rereview"

File: css/guests.css [hunk 11]
```diff
@@ -0,0 +1,3 @@
+.nav-icon-sharingguests {
+	background-image: url('../../../core/img/actions/share.svg');
```
Comment (PVince81) [inline css/guests.css]: Setting the class "icon-share" on your element + a few tweaks should be enough.

Unless you tried that already and it didn't work ?

**Comment by individual-it** [comment]
what about unit tests for this changes?

**Comment by cortho** [comment]
@individual-it good question. there never were unit test in the history of the guests app, see the raised issue https://github.com/owncloud/guests/issues/4 (note the issue number). I guess I could dive more deeply into the integration tests and test this PR that way. Would you think this would be sufficient?

@PVince81 do you regard this as a blocker for this PR?

File: css/guests.css (hunk 11 above)
Comment (cortho) [inline css/guests.css]: I'll find out ...

**Comment by individual-it** [comment]
@cortho I would argue that the developers should write unit tests firsts and then integration tests can be added on top
https://martinfowler.com/bliki/TestPyramid.html
E.g. integration and specially UI tests are much slower/expensive than unit tests and its much harder to test edge cases

**Comment by PVince81** [comment]
yeah, it's unfortunate that there were no unit tests in the first place in this project...

Would be good to add some. Tricky part is to add JS tests as the karma framework hasn't been imported yet. You could look at how it's done here https://github.com/owncloud/activity/tree/master/tests/js and here https://github.com/owncloud/activity/blob/master/.travis.yml#L76. Bonus points if you can make it a Makefile rule "test-js".

File: css/guests.css (hunk 11 above)
Comment (cortho) [inline css/guests.css]: Re-checked this. The class  name is more or less autogenerated and always starts with "nav-". There is no suitable css definition present  to handle this. Other apps (files_sharing, files_trashbin) have a custom css as well.

Conclusion: we'll leave this css.

**Comment by PVince81** [comment]
@cortho the most difficult part is likely to setup the JS tests from scratch. If you have trouble with this please let me know and I'll setup this myself and then you can focus on just writing JS tests ?

**Comment by PVince81** [comment]
I'd be fine merging this now and adding tests later on...

**Comment by PVince81** [comment]
after the QA phase perhaps

**Comment by IljaN** [comment]
@cortho merge?

**Comment by PVince81** [comment]
depends if this is supposed to be in the current release, because it would require additional QA then: @owncloud/qa

**Comment by cortho** [comment]
@IljaN @PVince81 originally this was not planned for the current release, see https://github.com/owncloud/guests/issues/128, so it was not included in the current QA. Regarding the tests, this week I cannot spend time on it.

**Comment by davitol** [comment]
I think i was told by @pmaier1 it won't be included in this release

**Comment by PVince81** [comment]
@cortho can you rebase and solve the conflicts ?

**Comment by cortho** [comment]
@PVince81 I was on holiday and yes, I will rebase.

**Comment by PVince81** [comment]
PR to bring in karma first: https://github.com/owncloud/guests/pull/186

**Comment by PVince81** [comment]
need to schedule time for adding unit tests.

Estimate: 0.25-0.5md

**Comment by PVince81** [comment]
Superseded by https://github.com/owncloud/guests/pull/214, had to do it to be able to push on local branch and later rebase
